### ISSUE TITLE
Need `string_keyed_label_list_dict`

### ISSUE THREAD

**Issue by excitoon** [bazelbuild/bazel#7989]
Motivation.
1) Have an ability to write `my_pkg_tar` not repeating directories list:
```
my_pkg_tar(
  name='xxx'
  files={
    'bin': [ *loooong_list ], # * just for clarity
    'lib': [ *one_more_long_list ],
  }
)
```
instead of:
```
my_pkg_tar(
  name='xxx'
  files={
    ':aaa': 'bin',
    ':aab': 'bin',
    ':aac': 'bin',
    ':aad': 'bin',
    ':aae': 'bin',
    ':aaf': 'bin',
    ':aag': 'bin',
......... looooooong list
    ':zzs': 'lib',
    ':zzt': 'lib',
    ':zzu': 'lib',
    ':zzv': 'lib',
    ':zzw': 'lib',
    ':zzx': 'lib',
    ':zzy': 'lib',
    ':zzz': 'lib',
  }
)
```
2) Dicts can not be added to `select`'s of dicts (and even more `+` is deprecated), one can not write:
```
{ ':aaa' : 'bin' } + select(....) + { ':zzz': 'lib' }
```

And we could write:
```
{ 'bin' : some_list + select(...), 'lib' : some_other_list }
```

**Comment by brandjon** [comment]
See [this explanation](https://github.com/bazelbuild/bazel/blob/01d1334978db34a826a624938c8ae8e69a3cb37f/src/main/java/com/google/devtools/build/lib/packages/Type.java#L58-L80) of why we're loathe to add new attribute types.

**Comment by Wyverald** [comment]
I'm also of the opinion that `string_keyed_label_dict` would be very useful, and wouldn't be very disruptive to our attribute-handling machinery (e.g. it's very intuitive to think about how to represent these in `query`). It's a rather annoying structure to represent using existing attribute types; presumably, one would use a `label_list` with a bunch of targets each of which would serve as an entry in the dict, so a `<string, label>` tuple -- which makes the representation much more verbose. Additionally, this "deconstruction" is even more awkward in a module extension API.

With that said, `string_keyed_label_list_dict` seems less generally useful to me. It could fairly easily be deconstructed into `{string: label} + targets with [label]` (for BUILD) or `{string: string} + tags with [label]` (for module extensions), and this deconstruction doesn't blow up the representation since the "label list" part would presumably take up most of the space.

cc @fmeum

**Comment by lberki** [comment]
3ef2abc implements the string-to-label attribute type. 

For complicated reasons, string-to-label-list-dict would be considerably more complicated to implement and as the last comment by @Wyverald says, the utility of string-to-label-list-dict over string-to-label-dict is marginal. So I consider this bug fixed; on the off chance that string-to-label-list-dict is indispensable, please re-open, but don't expect a very quick resolution.

**Comment by fmeum** [comment]
Could we backport this to 7.4.0 and 8.0.0? Otherwise it will take a long time for this to be adoptable by rulesets.

**Comment by lberki** [comment]
Yeah, I will need to do some backporting (not just this change), but I'll also need to earmark some time for it...

**Comment by lberki** [comment]
https://github.com/bazelbuild/bazel/pull/23997 for 8.0.0

**Comment by fmeum** [comment]
@lberki Now that I think of it, aren't we going to add a string-to-label-list type for https://github.com/bazelbuild/bazel/issues/23802?

**Comment by lberki** [comment]
@fmeum I don't know, I was hoping @katre would decide that -- it's just that I didn't want to commit to it on its own merits because it's way too much work for the little improvement over string-to-label-dict.

**Comment by katre** [comment]
Yes, string-to-label-list would be very useful for #23802. I;m afraid I don't understand @Wyverald's comment because I am not familiar with the internal representation of attribute typed.

**Comment by lberki** [comment]
@Wyverald's comment says that you don't usually need a string-to-label-list-dict because you can transform that into a string-to-label-dict and a a few separate rules like this:

```
rule(
  name = "main", 
  attr = {
    "1": ["//one", "//eins", "//egy"],
    "2" :["//two", "//zwei", "//ketto"]})
```

can be transformed into

```
rule(
  name = "main",
  attrs = {
    "1": ":un",
    "2": ":deux"]}

newrule(name = "un", numbers = [":one", ":eins", ":egy"])
newrule(name = "deux", numbers = [":two", ":zwei", ":ketto"])
```

**Comment by katre** [comment]
Yes, it's possible to do that, but it's not ergonomic, and requires an extra dependency lookup.

The proposed use case from #23802 is for adding new exec constraints to exec groups. The specific proposal is for the syntax:
```
some_target(
    ...,
    exec_compatible_with = {
        "": ["//:has_fast_cpu", "@platforms//:cpu:x86_64"],
        "test": ["//:has_gpu"],
    },
)
```

So a) any target can use this attribute, and b) it's part of toolchain resolution, and thus happens before any dependencies are resolved.

Due to this, trying to decompose into a separate `constraint_value_group` rule just for bundling adds new dependency edges and nodes and slows down `ConfiguredTargetFunction` (by adding at least one new skyframe restart to resolve these). While this won't be a commonly-used feature, it is a requested feature, and we should try to make it as simple as possible for users, even if it makes for some complexity in the attribute parsing process.

**Comment by lberki** [comment]
I won't volunteer for adding `string_keyed_label_list_dict`, but I did do some research and it looks like it's much less work to do so than I had originally thought.

**Comment by purkhusid** [comment]
Can this issue be reopened since this was not implemented?

**Comment by katre** [comment]
Re-opened, not sure if I'm going to have time to tackle it, however.

**Comment by katre** [comment]
Good news, being fixed as part of #24964.

**Comment by fmeum** [comment]
@katre The native attribute type has been added, but it isn't exposed to Starlark yet. Half the work is done though.

**Comment by katre** [comment]
Oh right, thanks for clarifying that.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 9.0.0 RC2](https://github.com/bazelbuild/bazel/releases/tag/9.0.0rc2). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=9.0.0rc2. Thanks!
